src/LanguageServer/Impl/Indexing/IndexParser.cs

```diff
@@ -37,34 +41,64 @@ internal sealed class IndexParser : IIndexParser {
             _version = version;
         }
 
-        public void Dispose() {
-            _allProcessingCts.Cancel();
-            _allProcessingCts.Dispose();
+        public Task ParseAsync(string path, CancellationToken cancellationToken = default) {
+            lock (_syncObj) {
+                CancelCurrentParse();
+                _tcs = new TaskCompletionSource<object>();
+                _linkedParseCts = CancellationTokenSource.CreateLinkedTokenSource(_allProcessingCts.Token, cancellationToken);
+                _parseTask = Task.Run(() => Parse(path, _linkedParseCts));
+                return _tcs.Task;
```

Comment:
Why not just `Task.Run(() => Parse(path, _linkedParseCts))` ? TaskCompletionSource is used when there are two separate entities participate in the process. First one, that creates the TCS and starts the async task, does not or cannot await. Second entity may come later and ask for the result. The latter does not initiate the task. `Task ParseAsync` starts the parse but it returns a task so it is awaitable - which contradicts (1). Is there second entry point that just receives the TCS?